### PR TITLE
fix: add subscription right after observer creation in react-query

### PR DESCRIPTION
**Pull request by artem-malko**
This PR attempts to close https://github.com/TanStack/query/issues/5443

### Opened questions

* It is really hard to write any tests for such problem(
* Is it clear about useLazyRef? Looks like a hack

Please, read the [issue](https://github.com/TanStack/query/issues/5443) before code review.

@TkDodo and @ericbiewener you have the full context)

### CODE REVIEW

**Comment by artem-malko** [comment]
@TkDodo if there is a way to check react-query in my private project from that pull request, I'd like to use it)

File: packages/react-query/src/useLazyRef.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  const ref = React.useRef<Value | typeof none>(none)
+
+  // if it's not initialized (1st render)
+  if (ref.current === none) {
+    // we initialize it
+    ref.current = initializer()
+  }
+  // now we return the initialized ref
+  // "as" is used to not leak any info about type of "none" outside of useLazyRef
+  return ref as React.MutableRefObject<Value>
```
Comment (TkDodo) [inline packages/react-query/src/useLazyRef.ts:25]: I don't understand why this is preferred to `useState` ? It just seems like a more complicated, harder-to-type way of achieving the exact same thing?

File: packages/react-query/src/useBaseQuery.ts [hunk 2]
```diff
@@ -65,37 +66,75 @@ export function useBaseQuery<
...
+      TQueryFnData,
+      TError,
+      TData,
+      TQueryData,
+      TQueryKey
+    >(queryClient, defaultedOptions)
+
+    initiallyUnsubscribe.current = observer.subscribe(() => {
+      needsToFlushStoreChange.current = true
+    })
```
Comment (TkDodo) [inline packages/react-query/src/useBaseQuery.ts:90]: I thought the agreement was to use layout effect instead ? I'm pretty sure it's against the rules of react to create a subscription inside an initializer, which should be pure.

**Review by TkDodo — CHANGES_REQUESTED** [review]

**Comment by TkDodo** [comment]
superseded by:
- https://github.com/TanStack/query/pull/5474

File: packages/react-query/src/useBaseQuery.ts (hunk 2 above)
Comment (artem-malko) [reply packages/react-query/src/useBaseQuery.ts:90]: It was an attempt to create a subscription on the first render) But yeah, you're right, it's quite tricky

**Pull request closed without merging**